src/datahike/index.cljc

```diff
@@ -0,0 +1,108 @@
+(ns datahike.index
+  (:require [datahike.index.hitchhiker-tree :as dih]
+            [datahike.index.persistent-set :as dip])
+  #?(:clj (:import [hitchhiker.tree.core DataNode IndexNode]
+                   [me.tonsky.persistent_sorted_set PersistentSortedSet])))
+
+;; TODO add doc to each function
+(defprotocol IIndex
+  (-all [index])
+  (-seq [index])
+  (-count [index])
+  (-insert [index datom index-type])
+  (-remove [index datom index-type])
+  (-slice [index from to index-type])
+  (-flush [index backend])
+  (-transient [index])
+  (-persistent! [index]))
+
+
+(extend-type DataNode
```

Comment:
This is not related to an issue with the code but more something that I would like to understand. What is the difference between `DataNode`and `IndexNode`?

Reply:
In B+-trees IndexNodes only manage branching and all the data is at leaves (DataNodes).